File: src/memory.zig
const MAX_MEMORY_SIZE: i32 = 4096;

const Memory = struct {
    data: u8[MAX_MEMORY_SIZE],
    pub fn new() Memory {
        const self: Memory = .{};
        @memset(self.data, 0);
        return self;
    }

    pub fn write(self: *Memory, address: u12, data: u8) void {
        self.data[address] = data;
    }

    pub fn read(self: *Memory, address: u12) u8 {
        return self.data[address];
    }
};

File: memory/memory.go
package memory

import "github.com/AlanRostem/mu-8/mu8"

const MemorySize = 0x1000

type Memory struct {
	data [MemorySize]mu8.Byte
}

func NewMemory() *Memory {
	return &Memory{
		data: [MemorySize]mu8.Byte{},
	}
}

func (m *Memory) Write(addr mu8.Address, value mu8.Byte) {
	m.data[addr.Value()] = value
}

func (m *Memory) Read(addr mu8.Address) mu8.Byte {
	return m.data[addr.Value()]
}
